File: frontend/src/grid-view/kpi.ts
import { num, uiLocale } from "./format";
import type { KpiSpecDict, ResolvedKpiDict, RowDict } from "./types";

interface KpiCardInput extends ResolvedKpiDict {
  value_fmt?: string;
}

export function formatKpiValue(
  value: number | string,
  fmt: string | undefined,
): string {
  var n = Number(value);
  if (!Number.isFinite(n)) return String(value);
  fmt = fmt || "number";
  if (fmt === "currency") {
    return n.toLocaleString(uiLocale(), { maximumFractionDigits: 0 });
  }
  if (fmt === "percent") {
    return n.toFixed(1) + "%";
  }
  if (fmt === "number") {
    if (Math.abs(n - Math.round(n)) < 1e-9) {
      return Math.round(n).toLocaleString(uiLocale());
    }
    return n.toLocaleString(uiLocale(), {
      minimumFractionDigits: 2,
      maximumFractionDigits: 2,
    });
  }
  return String(value);
}
export function aggregateKpi(spec: KpiSpecDict, rows: RowDict[]): number {
  const agg = spec.aggregate || "count";
  if (agg === "count") return rows.length;
  const key = spec.columnKey ?? spec.column_key ?? spec.field;
  if (!key) return 0;
  const nums: number[] = [];
  rows.forEach((row) => {
    const parsed = num(row[key]);
    if (parsed !== null) nums.push(parsed);
  });
  if (agg === "sum")
    return nums.reduce(function (a, b) {
      return a + b;
    }, 0);
  if (agg === "avg")
    return nums.length
      ? nums.reduce(function (a, b) {
          return a + b;
        }, 0) / nums.length
      : 0;
  if (agg === "min") return nums.length ? Math.min.apply(null, nums) : 0;
  if (agg === "max") return nums.length ? Math.max.apply(null, nums) : 0;
  return 0;
}
export function resolveKpis(
  specs: KpiSpecDict[],
  rows: RowDict[],
): ResolvedKpiDict[] {
  return (specs || []).map(function (spec) {
    var raw = aggregateKpi(spec, rows);
    return {
      label: spec.label || "",
      valueFmt: formatKpiValue(raw, spec.format),
      rawValue: raw,
      tone: spec.tone || "default",
      icon: spec.icon || undefined,
    };
  });
}
export function kpiCardHtml(kpi: KpiCardInput): string {
  var icon = kpi.icon || "\uD83D\uDCCA";
  var label = kpi.label || "";
  var value = kpi.valueFmt || kpi.value_fmt || "";
  return (
    '<span class="cm-kpi-icon" aria-hidden="true">' +
    icon +
    "</span>" +
    '<div class="cm-kpi-body">' +
    '<span class="cm-kpi-label">' +
    label +
    "</span>" +
    '<span class="cm-kpi-value">' +
    value +
    "</span>" +
    "</div>"
  );
}
export function initKpiStrip(
  root: Element,
  kpis: ResolvedKpiDict[],
  columns: number | undefined,
): void {
  if (!root || !kpis?.length) return;
  root.innerHTML = "";
  root.className = `cm-kpi-grid cm-kpi-cols-${columns || 4}`;
  kpis.forEach((kpi) => {
    const card = document.createElement("div");
    card.className = `cm-kpi-card cm-kpi-tone-${kpi.tone || "default"}`;
    card.innerHTML = kpiCardHtml(kpi);
    root.appendChild(card);
  });
}
export function initAllKpi(scope?: Document | Element): void {
  const root: Document | Element = scope || document;
  root.querySelectorAll("[data-cm-kpi-config]").forEach((node) => {
    if (!(node instanceof HTMLElement)) return;
    if (node.dataset.cmKpiReady) return;
    const kpis: ResolvedKpiDict[] = JSON.parse(
      node.dataset.cmKpiConfig || "[]",
    );
    const columns = parseInt(node.dataset.cmKpiColumns || "4", 10);
    initKpiStrip(node, kpis, columns);
    node.dataset.cmKpiReady = "1";
  });
}
export const Kpi = { initKpiStrip, initAllKpi, kpiCardHtml };

File: src/django_grid_view/render/kpi.py
from __future__ import annotations

from collections.abc import Sequence

from django_grid_view.render.format import format_value
from django_grid_view.types.artifact import ResolvedKpi
from django_grid_view.types.enums import KpiAggregate
from django_grid_view.types.json import RowDict
from django_grid_view.types.kpis import KpiSpec
from django_grid_view.types.numbers import parse_number


def resolve_kpis(specs: Sequence[KpiSpec], rows: Sequence[RowDict]) -> tuple[ResolvedKpi, ...]:
    return tuple(_resolve_one(spec, rows) for spec in specs)


def _resolve_one(spec: KpiSpec, rows: Sequence[RowDict]) -> ResolvedKpi:
    raw = _aggregate(spec, rows)
    return ResolvedKpi(
        label=spec.label,
        value_fmt=format_value(raw, spec.format),
        raw_value=raw,
        tone=spec.tone.value,
        icon=spec.icon,
    )


def _aggregate(spec: KpiSpec, rows: Sequence[RowDict]) -> float | int:
    match spec.aggregate:
        case KpiAggregate.COUNT:
            return len(rows)
        case KpiAggregate.SUM:
            return _sum_rows(rows, spec.column_key)
        case KpiAggregate.AVG:
            values = _numeric_values(rows, spec.column_key)
            return sum(values) / len(values) if values else 0
        case KpiAggregate.MIN:
            values = _numeric_values(rows, spec.column_key)
            return min(values) if values else 0
        case KpiAggregate.MAX:
            values = _numeric_values(rows, spec.column_key)
            return max(values) if values else 0
    return 0


def _sum_rows(rows: Sequence[RowDict], column_key: str | None) -> float:
    if not column_key:
        return 0.0
    return sum(_numeric_values(rows, column_key))


def _numeric_values(rows: Sequence[RowDict], column_key: str | None) -> list[float]:
    if not column_key:
        return []
    values: list[float] = []
    for row in rows:
        parsed = parse_number(row.get(column_key))
        if parsed is not None:
            values.append(parsed)
    return values
